### DISCUSSION TITLE
Problem with Interval Trigger in Background Scheduler

### DISCUSSION THREAD

**Discussion by sri-dhurkesh** [agronholm/apscheduler#703] [Q&A]
I am currently working on Algorithmic trading and I am using an API to retrieve data every second. The API request takes approximately 300ms. However, I am experiencing an issue where the background scheduler with an interval trigger is skipping some time. Specifically, on average every 20 seconds, I receive a warning that states "` skipped: maximum number of running instances reached.`" I have tried using a different scheduler library, the schedule library, and it works without skipping any time. I suspect that this issue may be caused by thread overload, but I am not certain. Can anyone provide guidance on how to solve this problem? Below is the code for the scheduler that I am currently using:

`scheduler.add_job(id="Monitor", func=get_data, trigger="interval", seconds=1, coalesce=True, timezone="Asia/Kolkata")`

Note: I am not able to increase the max instance.Because it is a synchronous process.

**Comment by agronholm** [comment]
Have you tried setting a more lenient `misfire_grace_time`?

**Reply by agronholm** [reply]
Also, have you measured how much time the job actually takes? Are you sure that it can complete within 1 second?

**Reply by sri-dhurkesh** [reply]
Yeah @agronholm, But the problem, is I am not able to wait for the grace period if the function takes more than one second.

**Reply by agronholm** [reply]
What do you expect the scheduler to do then if the previous run is still ongoing when it's supposed to launch a new one?

**Reply by sri-dhurkesh** [reply]
> Also, have you measured how much time the job actually takes? Are you sure that it can complete within 1 second?

Yeah, I am damn sure about this. Just now tested the same function with the schedule library.

**Reply by agronholm** [reply]
Does the schedule library place any restrictions on how many instances of a scheduled task can be running at once? What I'm asking is whether you actually measured the time taken by the target function.

**Reply by agronholm** [reply]
According to the executor, the previous instance is still running, so without any code to try on my end, this is all I can say on the subject.

**Reply by sri-dhurkesh** [reply]
The purpose of the function is to retrieve the current price of a stock from an API every second and check if it is above a certain predefined price. If it is, the function will execute a sell order for the stock. 
In this scenario, I do not need the "misfire_grace_time"  as I want to always have the most current price, regardless of whether the function was able to execute at the exact second it was scheduled to. The function should run every second without any delay or skipping.

**Reply by agronholm** [reply]
You're not answering my questions. If the function cannot complete before the next scheduled run, you will get a warning like above.

**Reply by sri-dhurkesh** [reply]
> Does the schedule library place any restrictions on how many instances of a scheduled task can be running at once? What I'm asking is whether you actually measured the time taken by the target function.

Yeah @agronholm , I actually measured the time taken to run the target function for every second using both apschduler and schedule library. In schedule library, it always takes around 200 ms.

**Reply by agronholm** [reply]
Ok, can you construct a trivial script that reproduces the issue, substituting `get_data()` with `sleep()` (with an argument of `0.2`)?

**Reply by sri-dhurkesh** [reply]
> You're not answering my questions. If the function cannot complete before the next scheduled run, you will get a warning like above.

I got like `skipped: maximum number of running instances reached.`. There no error other than this

**Reply by sri-dhurkesh** [reply]
> Ok, can you construct a trivial script that reproduces the issue, substituting `get_data()` with `sleep()` (with an argument of `0.2`)?

Okay, I'll create a script will post it ASAP.

**Reply by agronholm** [reply]
How's that going?
